index.js

```diff
@@ -848,81 +779,193 @@ function buildConstSerializer (location, input) {
 
-  if (typeof schema === 'boolean') {
-    return `json += JSON.stringify(${input})`
+  let mergedSchemaId = schema[mergedSchemaRef]
+  if (mergedSchemaId) {
+    const mergedLocation = getMergedLocation(context, mergedSchemaId)
+    return buildValue(context, mergedLocation, input)
   }
 
-  if (schema.$ref) {
-    location = resolveRef(context, location, schema.$ref)
-    schema = location.schema
+  mergedSchemaId = `__fjs_merged_${schemaIdCounter++}`
+  schema[mergedSchemaRef] = mergedSchemaId
+
+  const { allOf, ...schemaWithoutAllOf } = location.schema
+  const locations = [
+    new Location(
+      schemaWithoutAllOf,
+      location.schemaId,
+      location.jsonPointer
+    )
+  ]
+
+  const allOfsLocation = location.getPropertyLocation('allOf')
+  for (let i = 0; i < allOf.length; i++) {
+    locations.push(allOfsLocation.getPropertyLocation(i))
   }
 
-  if (schema.type === undefined) {
-    const inferredType = inferTypeByKeyword(schema)
-    if (inferredType) {
-      schema.type = inferredType
+  const mergedLocation = mergeLocations(context, mergedSchemaId, locations)
+  return buildValue(context, mergedLocation, input)
+}
+
+function buildOneOf (context, location, input) {
+  context.validatorSchemasIds.add(location.schemaId)
+
+  const schema = location.schema
+
+  const type = schema.anyOf ? 'anyOf' : 'oneOf'
+  const { [type]: oneOfs, ...schemaWithoutAnyOf } = location.schema
+
+  const locationWithoutOneOf = new Location(
+    schemaWithoutAnyOf,
+    location.schemaId,
+    location.jsonPointer
+  )
+  const oneOfsLocation = location.getPropertyLocation(type)
+
+  let code = ''
+
+  for (let index = 0; index < oneOfs.length; index++) {
+    const optionLocation = oneOfsLocation.getPropertyLocation(index)
+    const optionSchema = optionLocation.schema
+
+    let mergedSchemaId = optionSchema[mergedSchemaRef]
+    let mergedLocation = null
+    if (mergedSchemaId) {
+      mergedLocation = getMergedLocation(context, mergedSchemaId)
+    } else {
+      mergedSchemaId = `__fjs_merged_${schemaIdCounter++}`
+      optionSchema[mergedSchemaRef] = mergedSchemaId
+
+      mergedLocation = mergeLocations(context, mergedSchemaId, [
+        locationWithoutOneOf,
+        optionLocation
+      ])
     }
+
+    const nestedResult = buildValue(context, mergedLocation, input)
+    const schemaRef = optionLocation.getSchemaRef()
+    code += `
+      ${index === 0 ? 'if' : 'else if'}(validator.validate("${schemaRef}", ${input}))
+        ${nestedResult}
+    `
   }
 
-  if (schema.if && schema.then) {
-    return addIfThenElse(context, location, input)
+  let schemaRef = location.getSchemaRef()
+  if (schemaRef.startsWith(context.rootSchemaId)) {
+    schemaRef = schemaRef.replace(context.rootSchemaId, '')
   }
 
-  if (schema.allOf) {
-    mergeAllOfSchema(context, location, schema, clone(schema))
-    schema = location.schema
+  code += `
+    else throw new TypeError(\`The value of '${schemaRef}' does not match schema definition.\`)
```

Comment:
```suggestion
    else throw new TypeError(\`The value of '${schemaRef}' does not match the schema definition.\`)
```
?

Reply:
I didn't add this error. If you want to update the error message for this error, you can create another PR.